src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,575 @@
+     *
+     * @return a {@code KDF} object
+     *
+     * @throws NoSuchAlgorithmException
+     *     if no {@code Provider} supports a {@code KDFSpi} implementation for
+     *     the specified algorithm
+     * @throws InvalidAlgorithmParameterException
+     *     if the {@code AlgorithmParameterSpec} is an invalid value
+     * @throws NullPointerException
+     *     if the algorithm is {@code null}
+     */
+    public static KDF getInstance(String algorithm,
+                                  AlgorithmParameterSpec algParameterSpec,
+                                  Provider provider)
+        throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
+        Objects.requireNonNull(algorithm, "null algorithm name");
+        try {
+            Instance instance = GetInstance.getInstance("KDF", KDFSpi.class,
+                                                        algorithm,
+                                                        algParameterSpec,
+                                                        provider);
+            if (!JceSecurity.canUseProvider(instance.provider)) {
+                String msg = "JCE cannot authenticate the provider "
+                             + instance.provider.getName();
+                throw new SecurityException(msg);
+            }
+            return new KDF((KDFSpi) instance.impl, instance.provider, algorithm,
+                           algParameterSpec);
+
+        } catch (NoSuchAlgorithmException nsae) {
+            return handleException(nsae);
+        }
+    }
+
+    private static KDF handleException(NoSuchAlgorithmException e)
+        throws NoSuchAlgorithmException,
+               InvalidAlgorithmParameterException {
+        Throwable cause = e.getCause();
+        if (cause instanceof InvalidAlgorithmParameterException) {
+            throw (InvalidAlgorithmParameterException) cause;
+        }
+        throw e;
+    }
+
+    /**
+     * Derive a key, returned as a {@code SecretKey}.
+     * <p>
+     * The {@code deriveKey} method may be called multiple times once a
+     * {@code KDF} object is initialized.
+     * <p>
+     * Delayed provider selection is also supported such that the provider
+     * performing the derive is not selected until the method is called.
+     *
+     * @param alg
+     *     the algorithm of the resultant {@code SecretKey} object
+     * @param kdfParameterSpec
+     *     derivation parameters
+     *
+     * @return a {@code SecretKey} object corresponding to a key built from the
+     *     KDF output and according to the derivation parameters
+     *
+     * @throws InvalidParameterSpecException
+     *     if the information contained within the {@code KDFParameterSpec} is
+     *     invalid or incorrect for the type of key to be derived, or specifies
+     *     a type of output that is not a key (e.g. raw data)
+     *
+     */
+    public SecretKey deriveKey(String alg, KDFParameterSpec kdfParameterSpec)
```

Comment:
Are there cases where the parameters are correct, but the derive operation can still fail for other reasons? If so, I'm not sure we should be wrapping those in InvalidParameterSpecException. That exception should be thrown by the method initially when it inspects the parameters and before the derive operation begins.

This is why I mentioned previously if it makes sense to add a DerivationException class.

Reply:
First, very wrong parameters (say, null info, negative length) should not be create-able at all.

Then, in some cases, "correct" parameters could still be "invalid". For example, HKDF expand key length cannot exceed HashLen * 255, but HashLen is determined by the KDF algorithm. In this case, maybe an `InvalidAlgorithmParameterException` should be thrown because it does not conform to the spec.

Sometimes the implementation just does not support some parameters. For example, in PKCS #11 you cannot provide an arbitrary string as the algorithm name. Also, only if HKDF expand "info" is a well-known value that's recognized by the underlying implementation, `deriveData` is able to return a byte array. See 7a in https://docs.oasis-open.org/pkcs11/pkcs11-profiles/v3.1/os/pkcs11-profiles-v3.1-os.html#_Toc142307348. In these cases, maybe an `UnsupportedOperationException` should be thrown because the implementation does not support them.

Reply:
Let's leave it as is for now, but make a note to revisit this later.